resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -240,6 +240,9 @@ private[spark] class ApplicationMaster(
 
       logInfo("ApplicationAttemptId: " + appAttemptId)
 
+      // During shutdown, we may not be able to create an FileSystem object. So, pre-create here.
+      val stagingDirPath = new Path(System.getenv("SPARK_YARN_STAGING_DIR"))
+      val stagingDirFs = stagingDirPath.getFileSystem(yarnConf)
```

Comment:
The problem is we can not create a new filesystem object (this line ` val stagingDirFs = stagingDirPath.getFileSystem(yarnConf)`), if we are in shutdown and this is the first time we try to create a new filesystem object. We can not register a shutdown hook during a shutdown. That is why we should move the filesystem object creation outside of cleanupStagingDir itself. So, cleanupStagingDir called during shutdown won't try to create a filesystem object.

Reply:
Oh I see, it's in a hook. Is it safe to hold on to the FS object like this in the hook? maybe so, just makes me slightly uneasy about leaks

Reply:
Hi @srowen,

let me take a look at addressing the possible leak. Ideally when ShutdownManager executes each shutdownhook, it should automatically remove their references. But that does not seem to be the case right now. 

Mark this PR as WIP until we address/have a proposal for this concern.

Reply:
> Is it safe to hold on to the FS object like this in the hook?

I don't see why this would be an issue. It adds one additional reference to the state required for the hook, but this is negligible. FS objects are already globally cached, so in a typical case there is no new/additional FS instance being held onto by the hook; just a reference to the same FS object which would be used elsewhere throughout the driver. If `SPARK_YARN_STAGING_DIR` happens to be on a different file system than everything else, you'll end up with an additional FS reference, but most FS implementations (including HDFS) initialize all of their external connections lazily (upon the time of the first RPC) so you're just talking about a small amount of in-memory state.

> But that does not seem to be the case right now.

Can you elaborate @xinglin?

Reply:
the shutdownhook has a reference to the filesystem object, and Java won't GC/free an object until there is no more reference to it, right? 

In ShutdownHookManager.java, it adds all hooks into a Set (`hooks`). But in `executeShutdown()` to execute each hook, it does not remove each reference from the Set. So, the references should still be stored in `hooks`. To remove the reference, I think we should either add `hooks.clear()` at the end of `executeShutdown()` or remove the hook from Spark.

Reply:
After shutdown hooks complete, the JVM exits. There is no need to perform GC.

Reply:
got it. then I guess there is no concern here. Thanks,

I guess this PR is ready to be merged then.

Reply:
@srowen does this discussion address any concerns you had?